### ISSUE TITLE
Secrets fail to set the `uid`, `gid` and `mode` specified in `docker-compose.yml`

### ISSUE THREAD

**Issue by sanchitbapat** [docker/compose#9648]
**Description**

Docker secrets specified using the [long syntax](https://docs.docker.com/compose/compose-file/#long-syntax-3) for the `docker-compose.yml` file fail to set the specified `uid`, `gid` and `mode`. 

Also, from the docs, the default value of the `uid` and `gid` fields should be the user that runs the container however the value remains whatever was set on the host machine.

**Steps to reproduce the issue:**

1. Create a `Dockerfile`
```
❯ cat Dockerfile
FROM ubuntu:20.04

RUN adduser tester -u 1005 --disabled-password
CMD ["/bin/bash", "-c", "ls -al /run/secrets/target_secret_file.txt"]
```

2. Create a `docker-compose.yml`
```
❯ cat docker-compose.yml
services:
  secrets-tester:
    build: .
    secrets:
      - source: some_secret_file
        target: target_secret_file.txt
        uid: "1005"
        gid: "1005"
        mode: 0440

secrets:
  some_secret_file:
    file: somefile.txt
```
3. Create a text file for secrets (`somefile.txt`)
```
❯ cat somefile.txt
Text from a secret file
```

4. Run the service `docker compose run secrets-tester`

**Describe the results you received:**
Received Output:
```
❯ docker compose run secrets-tester
-rw-r--r-- 1 1000 1000 24 Jul 13 16:47 /run/secrets/target_secret_file.txt
```

**Describe the results you expected:**
Expected Output:
```
❯ docker compose run secrets-tester
-r--r----- 1 1005 1005 24 Jul 13 16:47 /run/secrets/target_secret_file.txt
```

**Additional information you deem important (e.g. issue happens only occasionally):**
Same behavior is observed in these cases:
1. Without creating the `tester` user
2. With same `source` and `target` names for the file in `docker-compose.yml`
3. Using the `uid` and `gid` for `root`
4. Using random values for `uid` and `gid`
5. Different values for `mode`

**Output of `docker compose version`:**

```
❯ docker --version
Docker version 20.10.17, build 100c701

❯ docker compose version
Docker Compose version v2.6.0
```

**Output of `docker info`:**

```
❯ docker info
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  app: Docker App (Docker Inc., v0.9.1-beta3)
  buildx: Docker Buildx (Docker Inc., v0.8.2-docker)
  compose: Docker Compose (Docker Inc., v2.6.0)
  scan: Docker Scan (Docker Inc., v0.17.0)

Server:
 Containers: 2
  Running: 0
  Paused: 0
  Stopped: 2
 Images: 1
 Server Version: 20.10.14
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Cgroup Version: 1
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: runc io.containerd.runc.v2 io.containerd.runtime.v1.linux
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 3df54a852345ae127d1fa3092b95168e4a88e2f8
 runc version: v1.1.2-0-ga916309
 init version: de40ad0
 Security Options:
  seccomp
   Profile: default
 Kernel Version: 5.10.102.1-microsoft-standard-WSL2
 Operating System: Ubuntu 20.04.4 LTS
 OSType: linux
 Architecture: x86_64
 CPUs: 8
 Total Memory: 7.763GiB
 Name: DESKTOP-IGS6AOM
 ID: MBVL:76QX:UWDQ:AK7Z:BNGW:PYGL:EXFG:ZHGD:JM53:RBFZ:4SQV:5I7Z
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false

WARNING: No blkio throttle.read_bps_device support
WARNING: No blkio throttle.write_bps_device support
WARNING: No blkio throttle.read_iops_device support
WARNING: No blkio throttle.write_iops_device support
```

**Additional environment details:**

**Comment by ev-adias** [comment]
Same thing happens to me. I'm trying to use secrets to workaround having to set the permissions on the host file, by setting the mode in the secret, but it end up with the exact same permissions of the host.

This issues comes up for me when using a keyfile for MongoDB replica authentication. The keyfile must not be world-readable, and in *nix this is easy to fix, in Windows (docker-desktop) it's not, so I was hoping that using secrets might fix the problem. It seems not

**Comment by markkrj** [comment]
As bind mounts do not support different permissions (uid, gid and mode), I suspect that in order to support this, the content of such secrets must be copied into the container (as in #9553), or, use POSIX ACLs (more complex). Am I right @ndeloof?
But why does the compose documentation mention such options if they are not supported in the first place?

**Comment by rijnhard** [comment]
well secrets and configs arent writeable (as per specification) so no reason why the implementation cant actually be a copy. 
unless the idea is that they can change during runtime...

**Comment by mario-gravel** [comment]
I'm using a secret for a RSA private key that must have a 0400 mode. Setting the uid, gid and mode is a requirement.

**Comment by ndeloof** [comment]
Same issue applies both to Compose v1 and v2
the root cause is that both use a bind mount to "inject" secret inside container, but bind mount doesn't allow a distinct UID/GID/mode to be exposed inside container.

A possible alternative I can imagine is to replace this approach with a `copy` (just like we do with support for secret set by `environment`). The sole drawback is that, doing so, changes to secret on host won't apply to secret inside container. Doesn't seem to be a major issue to me, but let me know if this would be a blocker for your usage.

**Comment by mario-gravel** [comment]
In my own opinion, having to rebuild the image to change a secret isn't an option. Major issue? Not now because we haven't reach the production, we're still developing. But once in production, it will be a security concern. 

The access to the container repository isn't as secure as the access to the production host. Somebody pulling the image will have access to the private keys. Not a good thing at all !

There must be control over the uid, gid and mode for a secret file.

**Comment by mario-gravel** [comment]
Does it implies a redeploy of the container or just the starting phase will be enough ?

It would be great if a refresh can be triggered manually (ex: docker compose refresh-secrets) to redo the copy in the container. Having a watchdog to do this automatically may consume resources unnecessary. Secrets don't change often but being able to update them without a restart can be a good solution.

**Comment by ndeloof** [comment]
we could copy secret again at any time, or have a dedicated command for this purpose, but the more obvious one would be to trigger a container `restart`

**Comment by mario-gravel** [comment]
As long as the secret file can be updated on the fly and still have the good `uid`, `gid` and `mode`, I'm good !

**Comment by rijnhard** [comment]
I agree with the suggestion on a copy implementation, my use case is local
development since I want to align local dev with a production swarm and a
quick restart on local is totally acceptable to me.

**Comment by cu** [comment]
I was testing out the secrets functionality using docker compose version 2.14.1 today and was surprised to learn that they don't work the way the docs say they should.

```
$ cat docker-compose.yaml 
services:
  bb:
    image: busybox
    command: sleep infinity
    secrets:
     - source: my_secret
       target: /secret_file_target
       uid: "1234"
       gid: "5678"
       mode: 0400

secrets:
  my_secret:
    file: ./secret_file_source
```
```
$ echo "this is my secret file" > secret_file_source
```
```
$ docker compose up -d
$ docker compose exec bb ls -l secret_file_target
-rw-rw-r--    1 1000     1000            23 Jan 11 22:38 secret_file_target
```
This also seems to affect `configs` as well.

**Comment by ndeloof** [comment]
Marking this issue as "kind/enhancement" as secrets never have supported having uid/gid set by compose (v1 or v2), this feature was introduced and only supported by docker swarm

**Comment by noahehall** [comment]
@ndeloof while they may be true

as a consumer we should be able to rely on the [documentation being the source of truth](https://docs.docker.com/compose/compose-file/#long-syntax-3)

```sh
Long syntax

The long syntax provides more granularity in how the secret is created within the service’s containers.

uid and gid: The numeric UID or GID that owns the file within /run/secrets/ in the service’s task containers. Default value is USER running container.
```

How does the *enhancement* tag affect the priority of seeing this issue resolved?

**Comment by AnderssonPeter** [comment]
It would be nice if this was supported when using `docker compose` I run the a container where the process runs as `nobody`.
So o give it access to the secret I have to give `nobody` read access to it on the host.

So if anyone gains access they can read the secrets in question without gaining root access.

**Comment by con-f-use** [comment]
> A possible alternative I can imagine is to replace this approach with a `copy` (just like we do with support for secret set by `environment`). The sole drawback is that, doing so, changes to secret on host won't apply to secret inside container. Doesn't seem to be a major issue to me, but let me know if this would be a blocker for your usage.

### IMPORTANT

**_A regular `COPY` directive for secrets is a bad idea, because the secret ends up in the image. The second someone accidentally uploads the image to a public docker repository, the secret is public knowledge._**

**Comment by ndeloof** [comment]
@con-f-use I'm not talking about a Dockerfile `COPY` but about using `docker cp` after the container has been created. It won't be stored inside an image

**Comment by ianhinder** [comment]
Will this allow us to use file-based secrets with a remote docker host, which is not currently possible with the bind-mount implementation?

**Comment by ndeloof** [comment]
@ianhinder yes indeed

**Comment by ianhinder** [comment]
Could this also be extended to configs?  Currently I have to build a new image with the config files baked in, as docker configs in compose don't work with a remote docker host.  If they were copied, similar to the secrets here, this would enable the use of local config files which appeared as standard configs on a remote host.

**Comment by con-f-use** [comment]
> as docker configs in compose don't work with a remote docker host

What are you refering to here, can you give an example?

**Comment by ndeloof** [comment]
using bind mount for local files indeed doesn't work with a remote docker host, and is a significant limitation for this use-case

**Comment by con-f-use** [comment]
Ah, now I get it, yes especially if you need the files on container startup, because then you can't `docker copy` or `scp` them.

**Comment by ndeloof** [comment]
actually you _can_ `docker cp` into a container before after it has been created and before it get started

**Comment by ianhinder** [comment]
Indeed - you can copy, but then your project configuration can't be expressed purely from a declarative compose file, and needs to be "booted" from a shell script or equivalent.  This makes it less easy to transfer it between environments.

**Comment by ndeloof** [comment]
sure, I'm not suggesting here to write your own hack-ish script to use `docker cp`, just saying we have an opportunity here to replace bind mounts for secrets/config with copy to better control permissions and ownership within container's filesystem

**Comment by ianhinder** [comment]
Exactly, my hope as well!

**Comment by ndeloof** [comment]
unfortunately this approach is currently blocked by https://github.com/moby/moby/issues/34142

**Comment by nicholastulach** [comment]
I assume the `cp` work around will _not_ work for build-time secrets…?

**Comment by ndeloof** [comment]
@nicholastulach exact, build time secrets are implemented by buildkit

**Comment by bodograumann** [comment]
> @peterge1998 the key thing about the docs is the explicit wording
> 
> `uid and gid: The numeric UID or GID that owns the file within /run/secrets/ in the service’s task containers. Default value is USER running container.`
> 
> a `service task` explictly refers to swarm containers

That was totally unclear to me. Can we make more explicit which features are only supported with docker swarm?

**Comment by marcpa00** [comment]
> @peterge1998 the key thing about the docs is the explicit wording
>
> uid and gid: The numeric UID or GID that owns the file within /run/secrets/ in the service’s task containers. Default value is USER running container.
>
> a service task explictly refers to swarm containers

@noahehall The docs quoted is present in [Services top-level element](https://docs.docker.com/compose/compose-file/05-services/) where clearly the word `services` refers to the children of `services` root element in the compose file, as stated in second paragraph of "Services top-level element" page.

Therefore, I would argue that `service's task` in that document naturally refers to a _service_ in a _compose file_ rather than a _swarm container_.

As @bodograumann says, if this is really a feature exclusive to docker swarm, it should be stated explicitly, or even removed from the document "Services top-level element" until it is actually something supported by `docker compose`.

**Comment by ndeloof** [comment]
@marcpa00 the compose specification is not intended to only cover docker compose features, that's fine _some_ feature only apply to an alternative tool (whenever docker swarm doesn't use the compose specification) as long as docker compose warn users about unsupported features. Also, docker compose _may_ be able to implement this feature once docker engine allows setting uid/gid on mounts, which has been discussed already. In the meantime, this is already available for environment-based secrets.

**Comment by nathanweeks** [comment]
> unfortunately this approach is currently blocked by [moby/moby#34142](https://github.com/moby/moby/issues/34142)

@ndeloof Would it be possible to utilize a tmpfs mount, which allows some control of the mountpoint ownership (https://github.com/docker/compose/issues/3425#issuecomment-423091370), in conjunction with the "volume with some initial content" idea that was mentioned for configs (https://github.com/docker/compose/issues/8707#issuecomment-1049919868)?

Somewhat of a naive question, as I'm not familiar with how Docker Compose is implemented at a systems level, or if a tmpfs would cause issues with scenarios like, e.g., host/container restarts.

**Comment by sabs21** [comment]
Any updates? I am running into this issue with Docker version 24.0.6, build ed223bc.
compose.yml
`
services:
  dev:
    secrets:
      - source: ssh_key
        mode: 0400
secrets:
  ssh_key:
    file: ~/.ssh/id_ed25519
`
Executing commands within container
`root@a0b9f2920f01:/usr/local/proj# ssh-add /run/secrets/ssh_key
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
@         WARNING: UNPROTECTED PRIVATE KEY FILE!          @
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
Permissions 0777 for '/run/secrets/ssh_key' are too open.
It is required that your private key files are NOT accessible by others.
This private key will be ignored.`

**Comment by stalb** [comment]
The permissions of the initial file are used in the container.
So changing the permissions of the `~/.ssh/id_ed25519` would do the trick.

The problem is when you need to mount the secret in two container which want non compatible permissions eg. 400 in one and 440 in the other.

**Comment by ndeloof** [comment]
latest docker compose release warn you about this feature not being supported
```
WARN[0000] secrets `uid`, `gid` and `mode` are not supported, they will be ignored 
```
the reason is that secrets are implemented as bind mounts, which comes with this limlitation.

**Comment by quarky42** [comment]
The documentation says mode, uid, and gid all work for compose services for both configs and secrets:
https://docs.docker.com/compose/compose-file/05-services/#long-syntax
and
https://docs.docker.com/compose/compose-file/05-services/#long-syntax-4

Why is this issue being closed as "not planned".  It is critical that files be loaded into the container with controlled / explicit permissions.   Go try running filebeat with a custom filebeat.yml file loaded in without setting the permissions on the file correctly and see how far you get.

What possible excuse can there be for this commitment to be backed out on?

Loading files into the service without control of permissions is absurd.

**Comment by ndeloof** [comment]
@quarky42 compose specification is a vendor-neutral definition of the compose file format, it doesn't define implementation details and limitation by Docker Compose.
This feature can't be implemented by Docker Compose as the docker engine has no support for config or secret (only when ran in swarm mode), and it uses a bind mount hack as a workaround, which can't offer such options.

**Comment by isujtauke** [comment]
> @quarky42 compose specification is a vendor-neutral definition of the compose file format, it doesn't define implementation details and limitation by Docker Compose. This feature can't be implemented by Docker Compose as the docker engine has no support for config or secret (only when ran in swarm mode), and it uses a bind mount hack as a workaround, which can't offer such options.

I feel like the confusion around this is warranted.  The documentation that one would use for building a docker-compose.yaml doesn't indicate anywhere that there's a split in how secrets work when you use docker compose vs swarm.  It's defined in the specification, and seemingly works partially in docker compose because of the aforementioned "bind mount hack"....though there's no indication via the documentation that any such hack is why it works in docker compose.

**Comment by ndeloof** [comment]
That's indeed a major challenge we have to address regarding inclusion of the Compose Specification in Docker docs, as then we miss "implementation details"

**Comment by sunnyphp** [comment]
> I'm using a secret for a SSH Private Key that must have a 0400 mode. Setting the uid, gid and mode is a requirement.
> 
> ```
>     secrets:
>       - source: ssh_host_ecdsa_key
>         target: /etc/ssh/ssh_host_ecdsa_key
>         mode: 0400
> ```
> 
> ```
> 2022-12-14 20:41:23 @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
> 2022-12-14 20:41:23 @         WARNING: UNPROTECTED PRIVATE KEY FILE!          @
> 2022-12-14 20:41:23 @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
> 2022-12-14 20:41:23 Permissions 0777 for '/etc/ssh/ssh_host_ed25519_key' are too open.
> 2022-12-14 20:41:23 It is required that your private key files are NOT accessible by others.
> 2022-12-14 20:41:23 This private key will be ignored.
> 2022-12-14 20:41:23 sshd: no hostkeys available -- exiting.
> ```

I'm found solution:
```
services:
  test:
    container_name: test
    image: busybox
    volumes:
      - ./source/.ssh:/root/ssh-copy
    tmpfs:
      - /root/.ssh
    command: sh -c "cp /root/ssh-copy/* /root/.ssh/ && chmod 0400 /root/.ssh/* && ls -la /root/.ssh"
```
Dirty hack, but acceptable for me.

**Comment by zorgoz** [comment]
Is there any roadmap for this feature (I needed this with configs, not secrets)? 
Just wondering: why is it not in the docs, that this is actually not supported?

**Comment by andrewvaughan** [comment]
@ndeloof 

> latest docker compose release warn you about this feature not being supported
> 
> ```
> WARN[0000] secrets `uid`, `gid` and `mode` are not supported, they will be ignored 
> ```
> 
> the reason is that secrets are implemented as bind mounts, which comes with this limlitation.

This error actually was more confusing than helpful to me, and is what led me to finding this discussion. It is seemingly counter to the [Compose documentation](https://docs.docker.com/compose/compose-file/05-services/#long-syntax-4), which makes it seem like they _are_ supported.

Only reading through this entire discucssion did I learn that the Docker team uses the term `service` as both a swarm-only term _and_ as a top-level mapping in the Compose documentation. There's no way users are just going to "know" that this is the case, and, at minimum, warrants an asterisk on the documentation explaining so. Requiring your users to understand how Swarm works should not be a dependency for being successful with Compose.

I'll add my $0.02 that this seems to be a severe limitation to Docker Compose Secrets and implementation should be reconsidered. This leaves the use of secrets at the whim of the image maintainer as to whether they build in a fashion that allows for access or not.

For example, in the formal Postgres docker container, they run all entrypoints as a `999:999` (`postgres`) user, meaning secrets become inaccessible during entrypoint and beyond. This severely limits the ability to extend that image's functionality with any use of secrets or add any dynamic capabilities through a Docker Compose configuration.

**Comment by spectrapulse** [comment]
I’m frustrated with the inadequate documentation and inconsistent support between Docker Swarm and Compose. This is the seventh instance I’ve encountered where Swarm lacks support for a feature that Compose offers, or vice versa.

It feels misleading to promote the capability of managing both environments with a single unified configuration in the documentation when, in reality, the latest version of Docker CE does not support the new Compose specification in Swarm. Unfortunately, Swarm lacks dedicated documentation that clearly delineates supported and unsupported features. This time, I’m struggling to mount secrets or configurations in a Compose environment because the application in the container expects specific file permissions, which necessitates a workaround. This issue likely stems from the decision to share a specification across two distinct tools with different levels of support for this spec, both of which are provided by the same vendor.

If the intention is for this to be a standalone specification, it should be renamed from “Compose File” to a more tool-agnostic term. The documentation should explicitly state when a feature is not supported by these tools—before developers encounter errors during the transition from development to a small production environment, not after, and certainly not through a GitHub Issue due to an unhelpful error message.

I can foresee further confusion with the introduction of Compose Bridge, particularly if it lacks comprehensive documentation. This oversight could lead to significant time wasted, especially in beta testing environments.

**Comment by ndeloof** [comment]
> It feels misleading to promote the capability of managing both environments with a single unified configuration

docker swarm does not rely on the compose specification, but on legacy compose file format. So there's no promise here you get a unified configuration.

> If the intention is for this to be a standalone specification, it should be renamed from “Compose File” to a more tool-agnostic term. 

"Compose specification" https://compose-spec.io/

>  further confusion with the introduction of Compose Bridge, particularly if it lacks comprehensive documentation

please be patient, Compose bridge is still fresh new experimental feature, more docs will come ASAP

**Comment by polarathene** [comment]
> latest docker compose release warn you about this feature not being supported
> 
> ```
> WARN[0000] secrets `uid`, `gid` and `mode` are not supported, they will be ignored 
> ```
> 
> the reason is that secrets are implemented as bind mounts, which comes with this limitation.

Just as additional awareness for others in this issue, you'll still get this warning for `environment` and `content` type sources for `secrets` / `configs` when you attempt to use those additional settings, but despite the warning [they are valid and applied for `environment` /  `content` sources](https://github.com/docker/compose/issues/12010), **not ignored**.

It probably would be beneficial to readers though if the docs at least clarified the common distinction of the types that aren't compatible/supported with `Docker Compose` considering the repo and docs hosted at Docker the majority of readers landing there are probably Docker Compose users?

The warning at least lets the reader know about the compatibility issue after they've parsed over the docs and attempted it... not the greatest UX but still something. It just shouldn't be adding to confusion though when the feature itself is supported by the configured source type (_or in the case of some config/secret sources being compatible in the `compose.yaml` while others are not, the warning could better clarify the incompatible source types if the official docs won't_).

**Comment by wz2b** [comment]
This capability is something you need even if you aren't using swarm.

**Comment by owl-with-beard** [comment]
I stumbled upon this issue exactly as @ev-adias did almost 2.5 **_years_** ago, wondering why my MongoDB key files were not working. I'm sorry, but this is clearly a flaw in the documentation. It's sad that this feature seems not to be on any road map to be brought to docker compose, but whatever. It is however almost rude and a gigantic waste of my time and the time of many other devs out there to not put anything in the documentation that just states that this is only implemented in Docker Swarm.

Best regards

**Comment by ndeloof** [comment]
>  It's sad that this feature seems not to be on any road map to be brought to docker compose

@owl-with-beard uid, gid and mode have been introduced bby docker swarm, and have no obvious equivalent in compose that would allow to add support on a non-swarm docker engine. The question is not about priorization in our roadmap, but feasability.

>  this is clearly a flaw in the documentation

indeed. The sad point here is that the compose specification (compose-spec.io) is supposed to be vendor neutral, so we can't bring much details on actual support by Docker Compose. I wonder we could relax this rule, as there's actually not much third-party contributions.

**Comment by quarky42** [comment]
This page is not updated to explain the limitation: 
https://docs.docker.com/reference/compose-file/services/#secrets

It needs to be. 

This next page was updated finally.  I have submitted feedback on docker pages asking that this be made clear because of the ridiculous amount of time I've spent on this issue previously. 
https://docs.docker.com/engine/swarm/secrets/#windows-support

I patently reject your defeatist attitude regarding support for a fundamental user, group, and mode control of files in compose services on Windows.   A 30 second search turns up an option for supporting uid, gid, and mode used WSL Meta data on Windows: 
https://learn.microsoft.com/en-us/windows/wsl/file-permissions

Docker runs on WSL by default.  If this feature was supported when WSL was available that would be of great benefit to many.

If for some reason that was prohibitive, I'm certain another solution could be found such as creating a ramdisk for secrets using tmpfs or other format, copying the secrets there at service startup, and applying the permissions. 

There are two ways to solve this off the top of my head.  Does a solution have to be perfect to at least be better than zero support? A solution with minor limitations is better than no solution at all when it comes to a hard requirement to support file permissions and ownership.  Multiple docker services require their config files to configured with permissions in order to boot.  The workarounds to avoid making this work in Docker Compose not only waste our time, they make divergent compose settings.  Being able to develop with docker compose and deploy with swarm seems to be the promise that Docker is falling short on.

**Comment by polarathene** [comment]
> It's sad that this feature seems not to be on any road map to be brought to docker compose, but so be it.

https://github.com/docker/compose/pull/12448#issuecomment-2665308130

This would resolve the lack of support apparently. It's unclear if that'll get merged though given the review feedback so far.

---

> I wonder we could relax this rule, as there's actually not much third-party contributions.

I think many that view the Compose reference docs at `docs.docker.com` are doing so with Docker Compose in mind.

You can always add an admonition or similar UI element to add some context for clarity. These types of caveats do harm the UX and cause confusion to users trying to leverage them when there's a lack of clarity and they need to troubleshoot and potentially find some resources elsewhere online clarifying the situation.

Another example I ran into recently was opt-out for a service from `docker compose up`, depending where you look there's `scale: 0` or `profiles: ["no-up"]` (or similar).

**Comment by ndeloof** [comment]
> This page is not updated to explain the limitation

should be, indeed.

> I patently reject your defeatist attitude

This is not defeatism, I've tried various approches to this issue in the past, without much success (most recently https://github.com/docker/compose/pull/12625). I'm not saying we will _never_ offer this feature, just we don't have the adequate features exposed by the engine API to succeed _yet_.

> A 30 second search turns up an option for supporting uid, gid, and mode used WSL Meta data on Windows

Which demonstrates you're under-estimating the actual challenge we have to address here. The question is not to get this running from WSL within Docker Desktop, but to get this supported in the Docker engine API, which hardly will accept such a non-portable option that can't be supported on Linux.

> A solution with minor limitations is better than no solution

Considering Docker engine _without_ swarm enabled has **no** support for configs and secrets, the existing behavior _already_ is a solution with known limitations. The other option is for you to include configs/secrets in your image as a build step, maybe relying on `dockerfile_inline` inside your compose file if you rely on an existing image.

@polarathene https://github.com/docker/compose/pull/12448 coms with some limitations and breaking changes (which I _tried_ to address in https://github.com/docker/compose/pull/12625). Those are significant enough so that we don't want to silently switch from bind mount to such a solution without an explicit opt-in by compose file author, otherwise we will have to deal with regression reports (many users _rely_ on the bind mount behavior to get config/secret dynamically updated in container)

**Comment by ndeloof** [comment]
Just a quick note:

With lack of native support for secrets/configs in docker engine (non-swarm) a possible workaround is to build a specific image for compose to COPY required files inside container:

```yaml
services:
  app:
    secrets:
       - source: app_config
          target: /etc/
          uid: <UID>  # not supported
          gid: <GID>

secrets:
   app_config:
      file: ./app.cfg
```

can be replaced by:

```yaml
services:
  app:
    build:
      dockerfile_inline: |
        FROM myapp
        COPY --chown <uid>:<gid> ./app.cfg:/etc/
```

This is far from being perfect, but at least wanted to get this documented in this issue for those blocked by this limitation

**Comment by polarathene** [comment]
@ndeloof no worries! I appreciate all the time and effort put into figuring out a solution 😎 

Slightly-offtopic, but is there any blockers for ID mapped support?
- I think that may be limited to rootful containers, and may be dependent upon Docker Engine moving to containerd 2.0 (_I thought that was the case with v28 due to some PRs bumping containerd deps to 2.0, but it still seems to ship with 1.7.x_).
- It'd allow for bind mounts with UID/GID mapping AFAIK, not sure about mode changes though.

I did try to get it working with compose and volume mount options a while back but ran into some issue that presumably is a limitation with Docker Engine, it worked fine to create the equivalent mount on the host without docker involved though.

**Comment by schaubl** [comment]
Hi all,
Happy to see discussions ongoing on this topic.
<hr>

> https://github.com/docker/compose/pull/12448#issuecomment-2665308130
> This would resolve the lack of support apparently.

@polarathene: Yes it does resolve it for `file` based _configs_ and _secrets_. 
The ones based on `environment` and `content` already support setting the `uid`, `gid` and `mode`.
My PR resolve it by using the same implementation as for `environment` and `content` (copying the file in the container).
It doesn't support directories, only a single file (as for `environment`, `content` and swarm/stack).
<hr>

I don't have a very strong opinion on that but I don't think directories should be supported because I think it doesn't match the original concept behind _configs_ and _secrets_ and it also have some implementation issues like using `mode` with sub-directories (should they have the same permissions as the files ?) and there are probably a lot of cases where not all the files should have the same permissions. Also the doc always mentions "file" and not "files"/"directories" (starting with the param name itself).
However I do agree that in some cases being able to inject files and directories could be useful.
Maybe the solution for this point would be to have an extra parameter named like `path`/`files`/`directories` which would be specific to Docker Compose (on top of `file` which will keep similar behaviour as in swarm/stack).
<hr>

> @polarathene https://github.com/docker/compose/pull/12448 coms with some limitations and breaking changes (which I tried to address in https://github.com/docker/compose/pull/12625). Those are significant enough so that we don't want to silently switch from bind mount to such a solution without an explicit opt-in by compose file author, otherwise we will have to deal with regression reports (many users rely on the bind mount behavior to get config/secret dynamically updated in container)

@ndeloof: Would it be enough to add a big warning of the behaviour change (and the way to update compose files) in the release notes as you initially mentioned here https://github.com/docker/compose/issues/11867#issuecomment-2141310118 ?
I don't know how many users rely on the bind mount behaviour but I find strange to use configs/secrets for this when bind mounts are already available. The only use case I can see is if a user uses a _secret_ to share sensitive informations to the container but this is even worse because the user think there are extra security measures which is not the case.
In case what the user wants is a bind mount, there is a drop-in replacement (bind mounts) but unfortunately if we have to inject a file into a container when `docker compose` is executed from a remote machine, there is no alternative.

**Comment by polarathene** [comment]
> I don't know how many users rely on the bind mount behaviour but I find strange to use configs/secrets for this when bind mounts are already available.
> The only use case I can see is if a user uses a _secret_ to share sensitive informations to the container but this is even worse because the user think there are extra security measures which is not the case.

That happened for me when I initially came across it and thought from the docs (_at least back then, haven't checked the current docs for it_) that it was meant to be more secure, along with the `uid`/`gid`/`mode` features to ensure it is when providing a file.

It's only because I noticed in the container none of those settings actually applied outside of swarm that I realized it was effectively a misleading bind mount with different syntax added as a fallback convenience for swarm users (_actual justification for it IIRC_).

---

As for avoiding regression reports from existing users... I suppose until you're ready to make it a breaking change, one option is to go opt-in as suggested with something like `secure: true` as the opt-in attribute.

Then clearly document that the `secrets` feature on non-swarm is not secure by default for legacy reasons, and requires additional explicit opt-in to toggle on the secure feature, with the caveat you lose the ability to sync updates from the host? 🤷‍♂ 

Technically since `uid`/`gid`/`mode` aren't supported in the existing support and this is specific to behaviour on non-swarm usage...there's no reason those users should be configuring those given they don't do anything thus you might as well make any of those the opt-in trigger instead, nothing new to add then. Documentation just needs to be clear that those options will break the current functionality of being treated as a typical bind mount? 🤔

**Comment by ndeloof** [comment]
> I don't think directories should be supported because I think it doesn't match the original concept behind configs and secret

A common pattern is to declare a whole `config.d` folder as a compose `config` and use it inside container. Many tools support this modularity pattern so that user don't hav to maintain a huge config file but can provide fragments

> I don't know how many users rely on the bind mount behaviour but I find strange to use configs/secrets for this when bind mounts are already available

This was introduced to offer _some_ parity with docker swarm, then users started relying on it in development to get immediate update on config changes (which could be implemented with `watch` nowadays). Then, with thousands example and stackoverflow suggestions you'll see tons of users to select this option without any specific reason. Still we have to manage our legacy to prevent compatibility breaks.

**Comment by schaubl** [comment]
@ndeloof thanks for your answers.
<hr>

To manage the legacy, I think it could be possible to create an environment variable like `USE_OLD_CONFIGS_SECRETS` which will default to `True`/`1` (similarly to the `COMPOSE_BAKE` introduced recently in [v2.33.0](https://github.com/docker/compose/releases/tag/v2.33.0)). 
At the same time, a warning can be added in the release notes to mention that the default value of this variable will, in the future, be changed to `False`/`0`.
With this, the current implementation will still be applied by default (for now) while allowing users to use the new implementation (by setting `USE_OLD_CONFIGS_SECRETS` to `False`) and for users relying on the current behaviour, they will have time to either migrate to bind mounts or to set `USE_OLD_CONFIGS_SECRETS` to `True`.
@ndeloof What do you think about this strategy ?
<hr>

Regarding the PR #12625, if I understood properly, bind mounts will still be used for `file` except if one of the `uid`, `gid` or `mode` params are set.
I have the impression that this could adds more confusion because the behaviour on handling the `file` will be completely different based on the usage or not of at least one of the 3 parameters mentioned above.
This is especially true when the `docker compose` command is executed from a remote machine. In this case, for example, if I want to inject my `/etc/hosts` file into the container, it will work if I explicitly set its `mode` but if I don't, it will take the `/etc/hosts` of the machine where the docker daemon is.
Also the behaviour can be different inside the same compose file, because the `file` param is in the top-level `configs` but `mode` is in the service-level `configs` so if I have a config used by two services where I need to change the permissions for only one service, the two different behaviour will be used and with my example with the `/etc/hosts` file, I will end up with two different file contents in the two containers even that they both use the same config.

```yaml
services:
  A:
    image: alpine
    configs:
      - source: etc_hosts
        target: /etc/hosts
    command: cat /etc/hosts
  B:
    image: alpine
    configs:
      - source: etc_hosts
        target: /etc/hosts
        mode: "0777"
    command: cat /etc/hosts
configs:
  etc_hosts:
    file: /etc/hosts
```

**Comment by ndeloof** [comment]
@schaubl fully agree. https://github.com/docker/compose/pull/12625 is just an experiment to prevent breaking changes, still unpleasant - also comes with multiple drawbacks I can't detail here. As you noticed, `secrets` support in compose was designed with local engine in mind (or Docker Desktop) but won't support an actual remote engine. On the other hand, if we use `copy` to inject files inside container, we then miss a mechanism to enforce consistency when local config/secret is updated.

The best option I have in mind is for docker engine to enable secrets/config support in standalone (non-swarm) mode. I can't tell about how complex this could be, but for sure the code is already there :P cc @colinhemmings 
@thaJeztah I haven't found an open issue requesting such a feature on moby repo, does this ring a bell ?

**Comment by GammaGames** [comment]
I found this thread after coming across this [in the docs](https://docs.docker.com/reference/compose-file/services/#long-syntax-4):

> Note that support for `uid`, `gid`, and `mode` attributes are not implemented in Docker Compose when the source of the secret is a [`file`](https://docs.docker.com/reference/compose-file/secrets/). This is because bind-mounts used under the hood don't allow uid remapping.
> 
> The following example sets the name of the `server-certificate` secret file to `server.cert` within the container, sets the mode to `0440` (group-readable), and sets the user and group to `103`. The value of `server-certificate` is set to the contents of the file `./server.cert`.
> 
> ```yaml
> services:
>   frontend:
>     image: example/webapp
>     secrets:
>       - source: server-certificate
>         target: server.cert
>         uid: "103"
>         gid: "103"
>         mode: 0o440
> secrets:
>   server-certificate:
>     file: ./server.cert
> ```

Am I reading this wrong, or does the documentation have and invalid example? 

Edit: Maybe because it's part of the spec and other programs support it

**Comment by ianhinder** [comment]
Yes, I think it's because it is part of the spec, but docker compose doesn't implement that part of the spec. And as far as I know, docker swarm mode doesn't implement ComposeV2 either, so I don't know that there is any implementation of this.
> The docker stack deploy command uses the legacy [Compose file version 3](https://docs.docker.com/reference/compose-file/legacy-versions/) format, used by Compose V1. The latest format, defined by the [Compose specification](https://docs.docker.com/reference/compose-file/) isn't compatible with the docker stack deploy command.

[ref](https://docs.docker.com/engine/swarm/stack-deploy/)

**Comment by vanyauhalin** [comment]
Please, add a note to this section that docker compose does not support this feature. Thank you.

https://docs.docker.com/reference/compose-file/services/#configs

**Comment by stefanfed** [comment]
What is the correct, secure way to have the contents of a local file readable only by root passed as a secret to an unprivileged container?

**Comment by stalb** [comment]
> What is the correct, secure way to have the contents of a local file readable only by root passed as a secret to an unprivileged container?

Either change the uid/gid/mode of the file which contains the secret or use an inline secret (cf https://docs.docker.com/reference/compose-file/secrets/#example-2).

When using inline secrets (as in this example: https://docs.docker.com/reference/compose-file/services/#long-syntax-5),  specifying uid/gid/mode works.
